iroh-dns-server/src/store/signed_packets.rs

```diff
@@ -1,20 +1,145 @@
-use std::{path::Path, sync::Arc};
+use std::{path::Path, result, time::Duration};
 
 use anyhow::{Context, Result};
 use iroh_metrics::inc;
 use pkarr::SignedPacket;
 use redb::{backends::InMemoryBackend, Database, ReadableTable, TableDefinition};
+use tokio::sync::{mpsc, oneshot};
+use tokio_util::sync::CancellationToken;
 use tracing::info;
 
 use crate::{metrics::Metrics, util::PublicKeyBytes};
 
 pub type SignedPacketsKey = [u8; 32];
 const SIGNED_PACKETS_TABLE: TableDefinition<&SignedPacketsKey, &[u8]> =
     TableDefinition::new("signed-packets-1");
+const MAX_BATCH_SIZE: usize = 1024 * 64;
+const MAX_BATCH_TIME: Duration = Duration::from_secs(1);
 
 #[derive(Debug)]
 pub struct SignedPacketStore {
-    db: Arc<Database>,
+    send: mpsc::Sender<Message>,
+    cancel: CancellationToken,
+    thread: Option<std::thread::JoinHandle<()>>,
+}
+
+impl Drop for SignedPacketStore {
+    fn drop(&mut self) {
+        // cancel the actor
+        self.cancel.cancel();
+        // join the thread. This is important so that Drop implementations that
+        // are called from the actor thread can complete before we return.
+        if let Some(thread) = self.thread.take() {
+            let _ = thread.join();
+        }
+    }
+}
+
+enum Message {
+    Upsert {
+        packet: SignedPacket,
+        res: oneshot::Sender<bool>,
+    },
+    Get {
+        key: PublicKeyBytes,
+        res: oneshot::Sender<Option<SignedPacket>>,
+    },
+    Remove {
+        key: PublicKeyBytes,
+        res: oneshot::Sender<bool>,
+    },
+}
+
+struct Actor {
+    db: Database,
+    recv: mpsc::Receiver<Message>,
+    cancel: CancellationToken,
+    max_batch_size: usize,
+    max_batch_time: Duration,
+}
+
+impl Actor {
+    async fn run(mut self) {
+        match self.run0().await {
+            Ok(()) => {}
+            Err(e) => {
+                self.cancel.cancel();
+                tracing::error!("packet store actor failed: {:?}", e);
+            }
+        }
+    }
+
+    async fn run0(&mut self) -> anyhow::Result<()> {
```

Comment:
The `run0` reads somewhat weird to me.
How about this:
- Move the outer `loop` from `run0` into `run`
- Rename `run0` to `run_one_batch`

Reply:
I often have this pattern. It is basically just to make sure that every unexpected error shuts down the entire loop with a nice error logging, while still being able to use just ? syntax.

Not quite sure why there isn't a try!{ } macro that allows you to do a block of code where you can use ? syntax inside... That would be nicer.